drivers: dac: Enable for STM32F0 and STM32G4 series

martinjaeger opened the pull request:
I added the `#include` of the LL driver headers for all STM32 MCUs that have a DAC. This allows to manually add the DAC-specific devicetree nodes in an out-of-tree board and use the Zephyr driver.

For the F0 and G4 series MCUs (where I have a Nucleo board) I added also the devicetree nodes and tested the DAC sample in hardware.

martinjaeger committed:
boards: nucleo_g431rb: Enable DAC driver

Also add this board to DAC driver sample

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
drivers: dac: Enable for STM32G4 series

Add DAC nodes to devicetree, include LL headers in soc.h and add
pinmux defines for DAC.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
drivers: dac: Enable for STM32F0 series

Add DAC nodes to devicetree, include LL headers in soc.h and add
pinmux defines for DAC.

Signed-off-by: Martin Jäger <martin@libre.solar>

ABOSTM reviewed (changes requested):
Need to fix buildkite issue (tests/drivers/dac/dac_loopback)
Otherwise LGTM

martinjaeger committed:
tests: drivers: dac_loopback: Fix dependencies

Use platform_allow instead of platform_exclude, as the test depends
on specific boards. A new board with DAC support should not be added
to this test automatically.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented:
> Need to fix buildkite issue (tests/drivers/dac/dac_loopback)
> Otherwise LGTM

Here we go. I should have waited for CI to finish before asking for review ;)

Rebased and added 1e74892 to fix the loopback test, everything else kept the same as before.

erwango commented:
@martinjaeger seems there is a conflict with recent changes. Can you have a rebase ?

martinjaeger commented:
Rebased and removed the loopback test fix again that was already implemented in other PR merged in between.

ABOSTM reviewed (approved):
LGTM

erwango commented on soc/arm/st_stm32/stm32f1/soc.h:
```diff
@@ -64,6 +64,10 @@
 #include <stm32f1xx_ll_adc.h>
 #endif
 
+#ifdef CONFIG_DAC_STM32
```
I'm not in favor of such broad changes.
This creates confusion in the end since we don't actually know if DAC driver works as is for F1 series.
I prefer we don't add blindly code for things that have not been verified.
Please restrict the changes to the part you've been able to test.

erwango reviewed (changes requested):
Thanks for the change, but to minimize confusion, please limit the changes to the code you've been able to test.

Also, for the F0 part, we miss the board part, is this because it was tested on a out of tree board ?

martinjaeger commented:
> Thanks for the change, but to minimize confusion, please limit the changes to the code you've been able to test.
> 
> Also, for the F0 part, we miss the board part, is this because it was tested on a out of tree board ?

Yes, tested the F0 part with an out-of-tree STM32F072-based board. I can also quickly add an additional Nucleo_F072 board definition, if you think that's necessary.

I will explain my situation to justify why I suggest to include the LL headers for all STM32 MCUs in soc.h. My stable firmware branch is based on Zephyr 2.3 release, which supports the DAC for STM32L0. Now I developed a very similar board based on STM32G4. I can easily add the devicetree nodes for the DAC to the board.dts, but it won't work because of the missing include in soc.h. For this reason I had to re-implement the driver manually using LL because I couldn't wait for the next Zephyr release.

It seems that the DAC peripherals and drivers are the same across all STM32 MCUs. But of course I haven't tested all of them (don't have all those Nucleo boards). So I suggest to add the devicetree nodes only to MCUs which are validated in hardware. In this way, nobody should be confused by a driver that can be easily enabled with `status = "okay"`, but doesn't work.

erwango commented:
> > Thanks for the change, but to minimize confusion, please limit the changes to the code you've been able to test.
> > Also, for the F0 part, we miss the board part, is this because it was tested on a out of tree board ?
> 
> Yes, tested the F0 part with an out-of-tree STM32F072-based board. I can also quickly add an additional Nucleo_F072 board definition, if you think that's necessary.

Having a board enabling the feature allows to do non regression in CI, so it's always a nice addition. However I won't push to add a new board specifically for that.

> I will explain my situation to justify why I suggest to include the LL headers for all STM32 MCUs in soc.h. My stable firmware branch is based on Zephyr 2.3 release, which supports the DAC for STM32L0. Now I developed a very similar board based on STM32G4. I can easily add the devicetree nodes for the DAC to the board.dts, but it won't work because of the missing include in soc.h. For this reason I had to re-implement the driver manually using LL because I couldn't wait for the next Zephyr release.

> It seems that the DAC peripherals and drivers are the same across all STM32 MCUs. But of course I haven't tested all of them (don't have all those Nucleo boards). So I suggest to add the devicetree nodes only to MCUs which are validated in hardware. In this way, nobody should be confused by a driver that can be easily enabled with `status = "okay"`, but doesn't work.

I understand that not having drivers working in all series could create some frustration. However there is a trade-off to do between:
-allowing pushing 'partial' developments 
-and blocking driver merge until they support all boards of all series.
Granularity has been set to the series level, when tested at least on one board, which makes it doable for a contributor to make a driver support progress, even if it cannot be tested on all series, nor all available boards of a series.
Also, having a clear line is also important in community support requests as it allows to make a difference between:
- what is available (and requests a bug if not working)
- what is not yet supported (and hence if someone reports it is not working this is a enhancement request, with understandably lower prio than a bug).

Out-of-tree is there to provide flexibility to users to extend support of available features to their own context. This will will require scalable effort whether it is:
- to extend support of an available driver to a board of a driver supported series
- to extend support of an available driver to a board of a driver non supported series
- to extend support of an non supported driver
...

As I understand, maybe there is more to do in making things flexible for out-of-tree users, can you have a look at https://github.com/zephyrproject-rtos/zephyr/pull/23884 and see if this is going in a direction that would make things easier to you?

martinjaeger commented:
@erwango Didn't know the linked PR, which would indeed solve the issue in my application.

So I went back to include the headers only for the tested MCU series in this PR and will continue the discussion in the other PR.

FRASTM reviewed (approved):
why ignoring the DAC feature 

> | DAC       | on-chip    | DAC controller                      |

in the  `boards/arm/nucleo_f091rc/doc/index.rst`

martinjaeger committed:
boards: nucleo_f091rc: Enable DAC driver

Also add this board to DAC driver sample

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented:
> why ignoring the DAC feature
> 
> > | DAC       | on-chip    | DAC controller                      |
> 
> in the `boards/arm/nucleo_f091rc/doc/index.rst`

@FRASTM Updated now (and flashed to Nucleo F072 for testing, as I don't have the F091 available).

erwango reviewed (changes requested):
@martinjaeger Don't we miss dac configured pins in pinmux.c ?
As I understand, this has no impact as configuration is analog which should be reset state, but it might be good to do it in order to avoid potential conflicts with other device configuration..

martinjaeger commented:
> @martinjaeger Don't we miss dac configured pins in pinmux.c ?
> As I understand, this has no impact as configuration is analog which should be reset state, but it might be good to do it in order to avoid potential conflicts with other device configuration..

Good spot, thanks. Added them.
